File: Problems/Hirarchical Feature Flag System/Go/feature_flag_system.go
package main

import "errors"

type FeatureFlagSystem struct {
	repository Repository
}

func NewFeatureFlagSystem(repository Repository) *FeatureFlagSystem {
	return &FeatureFlagSystem{
		repository: repository,
	}
}

func (fsys *FeatureFlagSystem) AddChannel(channelID string, features map[string]bool, parents []string) error {
	channel := fsys.repository.Get(channelID)
	if channel != nil {
		return errors.New("can not add, channel already exits")
	}
	channel = NewChannel(channelID, features, parents)
	fsys.repository.Add(channelID, channel)
	return nil
}

func (fsys *FeatureFlagSystem) DeleteChannel(channelID string) error {
	channel := fsys.repository.Get(channelID)
	if channel == nil {
		return errors.New("channel doesn't exits")
	}
	fsys.repository.Delete(channelID)
	return nil
}

func (fsys *FeatureFlagSystem) SetFeature(channelID string, featureFlag string, value bool) error {
	channel := fsys.repository.Get(channelID)
	if channel == nil {
		return errors.New("channel doesn't exits")
	}
	channel.SetFeature(featureFlag, value)
	return nil
}

func (fsys *FeatureFlagSystem) getFeatureFromParentIfNotDefined(id string, featureFlag string) (*bool, error) {
	channel := fsys.repository.Get(id)
	if channel == nil {
		return nil, errors.New("channel doesn't exits")
	}

	value, ok := channel.GetFeatures()[featureFlag]
	if ok {
		return &value, nil
	}

	for _, id := range channel.GetParents() {
		inheritedValue, err := fsys.getFeatureFromParentIfNotDefined(id, featureFlag)
		if err != nil {
			continue
		}
		if inheritedValue != nil {
			return inheritedValue, nil
		}
	}

	return nil, nil
}

func (fsys *FeatureFlagSystem) GetFeature(channelID string, featureFlag string) (*bool, error) {
	channel := fsys.repository.Get(channelID)
	if channel == nil {
		return nil, errors.New("channel doesn't exits")
	}

	value, ok := channel.GetFeatures()[featureFlag]
	if ok {
		return &value, nil
	}

	var inheritedValue bool

	for _, id := range channel.GetParents() {
		inheritedValue, err := fsys.getFeatureFromParentIfNotDefined(id, featureFlag)
		if err != nil {
			continue
		}
		if inheritedValue != nil {
			return inheritedValue, nil
		}
	}

	inheritedValue = false

	return &inheritedValue, nil
}

File: Problems/Hirarchical Feature Flag System/Python/feature_flag_system.py
from channel import Channel

class FeatureFlagSystem:
    def __init__(self, repository):
        self.repository = repository

    def add_channel(self, channel_id, features, parents):
        channel = self.repository.get(channel_id)

        if channel is not None:
            raise ValueError("Channel already exists")

        channel = Channel(channel_id, features, parents)

        self.repository.add(channel_id, channel)

    def delete_channel(self, channel_id):
        channel = self.repository.get(channel_id)

        if channel is None:
            raise ValueError("Channel doesn't exists")

        self.repository.delete(channel_id)

    def set_feature(self, channel_id, feature_flag, value):
        channel = self.repository.get(channel_id)

        if channel is None:
            raise ValueError("Channel doesn't exists")

        channel.set_feature(feature_flag, value)

    def find_in_the_parent_chain(self, id, feature_flag):
        channel = self.repository.get(id)

        if channel is None:
            raise ValueError("Channel doesn't exists")

        if feature_flag in channel.get_features():
            return channel.get_features()[feature_flag]

        for parent_id  in channel.get_parents():
            value = self.find_in_the_parent_chain(parent_id, feature_flag)
            if value is not None:
                return value

        return None

    def get_feature(self, channel_id, feature_flag):
        channel = self.repository.get(channel_id)

        if channel is None:
            raise ValueError("Channel doesn't exists")

        if feature_flag in channel.get_features():
            return channel.get_features()[feature_flag]

        for parent_id  in channel.get_parents():
            value = self.find_in_the_parent_chain(parent_id, feature_flag)
            if value is not None:
                return value

        return False